### ISSUE TITLE
increase some limits with command line option

### ISSUE THREAD

**Issue by fabiangreffrath** [chocolate-doom/chocolate-doom#427]
Dear fraggle,

I know this request is again stretching the rules and maybe goes bexond the scope of the project. But...

As it is now, Chocolate Doom supports emulation of (nearly?) every version of Doom executables available for DOS. This includes even unofficial ones like v1.91 -- but not Doom+. As such, it is possible to play all available official releases of Doom -- but not NRFTL/NERVE.WAD, because that would require raised visplanes and drawsegs limits.

Thus I'd like to suggest an additional command line parameter "-plus" that emulates the behaviour of the Doom+ executable by raising the static limits by the factors given here:
http://prboom-plus.sourceforge.net/doom-plus.features.html

Implementation wise I found Gez's suggestion here quite elegant:
http://www.doomworld.com/vb/showthread.php?s=&postid=1116989#post1116989
although even kb1's approach could work well.

Is this worth considering?
- Fabian

**Comment by fabiangreffrath** [comment]
Sorry to bring this issue up again, but I still believe that a "secret" option to increase the most critical limits, which could be enabled by default if the IWAD is detected to be one of the FreeDoom IWADs, would help to calm down the discussions over there a bit:

https://www.doomworld.com/vb/freedoom/91221-freedoom-should-be-limit-removing/

**Comment by AXDOOMER** [comment]
> I've been asked this in the past and have put thought into considering it, but I personally consider it a step too far away from the project's core philosophy. Changes like these also complicate the code, which I'd like to avoid. Basically this is a feature that's probably best for forks such as your own Crispy Doom.

Can you make the details of your thoughts available? How is it "too far away from the project's core philosophy" when other hacks such as longtics, novert, dehacked and vvHeretic were added to the project? Justifications could be added to the philosophy document. 

Is a hack judged by its popularity, its release date or by what it does? Isn't just a hack a hack? Why is it fine for Chocolate-Doom to not be able to play levels that were playable with the hacked vanilla engine because it doesn't have this hack?

**Comment by fragglet** [comment]
> Can you make the details of your thoughts available? How is it "too far away from the project's core philosophy" when other hacks such as longtics, novert, dehacked and vvHeretic were added to the project?

I'll admit that `-longtics` is a minor blip in the consistency of the project's philosophy. I've explained my reasoning a bit in my latest comment on #806. I think it's a feature worth keeping but I think we should probably clamp down on it; for example, it shouldn't be possible to embed `-longtics` recordings inside a PWAD file, which I think is currently possible.

The vvHeretic change (aa3e8e11e588c35eec44ea58f1a9d06b02e3efe5) I think is very good to bring up. The way I rationalized this (#432) is that it's analogous to `-longtics`. I think we should change both this and the Doom `-longtics` code to forbid these extensions in PWAD files.

Here's how I see things: shorttics demos degrade mouse turning quality a lot. To see what I mean, try mouse movement with and without the `-record` parameter. The difference is subtle but it does affect the gameplay. So I think there's a lot of value in providing people the ability to "perfectly" record and play back Vanilla gameplay. But we should make sure that's all people are doing.

As for novert and dehacked, these are already explicitly covered in the philosophy document. Quote:

> Chocolate Doom includes some features that aren’t part of Vanilla Doom but exist for compatibility with DOS tools that interact with it. These are considered part of the Vanilla experience and ought to be treated as such. Some examples are: (novert and dehacked)

**Comment by chungy** [comment]
> it shouldn't be possible to embed -longtics recordings inside a PWAD file, which I think is currently possible.

This sounds reasonable. I can imagine somebody accidentally recording a longtics demo for a WAD and believing that everything is functioning properly in vanilla because it appears to work in Chocolate. I don’t know if that’s ever happened, but defaulting to vanilla-only playback should be fine. Edge case: combine `-longtics` without `-playdemo` or with `-playdemo demo1`: should this still disallow embedded longtics demos? I think the answer would be “no,” but I don’t have a strong opinion about it.

**Comment by AXDOOMER** [comment]
>  it shouldn't be possible to embed -longtics recordings inside a PWAD file, which I think is currently possible.

> This sounds reasonable. 

You must do the same with 1.666 demos for consistency. Seems like this decision was taken for user convenience rater than emulating vanilla Doom. By default, Chocolate-Doom should be emulating vanilla (defined as "DOS Doom 1.9" in the philosophy document), not 1.666 or the longtics hack for user convenience, because it's taking decisions without the user's consent (a user who knows CD is supposed to emulate DOS Doom 1.9, but doesn't know that CD performs this little trick without his knowledge) and by doing this, Chocolate-Doom is diverging from its goal.

**Comment by jmtd** [comment]
> Sorry to bring this issue up again, but I still believe that a "secret" option to increase the most critical limits, which could be enabled by default if the IWAD is detected to be one of the FreeDoom IWADs, would help to calm down the discussions over there a bit:

Personally, I'd recommend just ignoring the freedoom project, at least from a chocolate perspective. I was surprised when ports (inc chocolate) started adding support for finding the freedoom IWADs, and given the current state of freedoom inaction I'd rather not see changes to follow a short-term direction that could end up being reversed anyway.

**Comment by chungy** [comment]
> You must do the same with 1.666 demos for consistency. Seems like this decision was taken for user convenience rater than emulating vanilla Doom.

Agreed, but I believe Chocolate Doom already will not play back 1.666 demos without an explicit `-gameversion` passed on. I wonder if it'd be alright to automatically set the gameversion depending on what the DEMO1 header in the IWAD says... or even just a hash table of known IWADs with 1.9 as a fallback.

On the importance of Doom-Plus: Okay, I'll admit my timeline and popularity arguments were poor, let's forget them and consider Doom-Plus as a utility function. As fraggle was mentioning about `-longtics`, it significantly changes game behavior in a way that would be beneficial to players. Speedrunners appreciate the game controlling exactly the same whether recording a demo or not.

Doom-Plus has two apparent purposes: Increase the save game limit, and raise other static limits so a subset of limit-removing maps will actually run in the (otherwise) vanilla engine. The save game limit can already be optionally removed in Chocolate Doom, so we needn't concern about it. As for the static limits, I am not aware of any mappers that actually target Doom-Plus's expanded limits, instead it's either vanilla or limit-removing. Trying to include these expanded limits just seems like an exercise in madness. What if your favorite map uses 1060 visplanes on your favorite level? Doom-Plus will still crash, but is it much of a stretch to just expand Chocolate Doom's limits even further? What if Doom-Plus itself updates for expanded limits and we have to follow suit?

In my opinion, Crispy Doom is already a good extension of Chocolate Doom to support limit-removing maps, in a fashion actually appropriate for such maps: by removing the limits. Why not use Crispy Doom for these? It's quite a good fork.

Speaking of where to draw the line for vanilla hacks... someone just skilled enough can do quite a lot with only the binary of vanilla Doom to work on. Quasar had noted on IRC that x86 code is still x86 on any OS, and someone could probably keep 95% of it intact and have a "pure" vanilla port on Windows or Linux. This has happened with Transport Tycoon Deluxe, getting unofficial ports from DOS to Windows, getting massive resolution increases, added features, extended maps, with absolutely nothing available except the DOS binary. It even ended up getting reverse engineered and rewritten in C++. Where would we draw the line? :)

longtics makes the line blurry, but I think a handful of convenience features like it don't do any harm to the philosophy and target of Chocolate Doom. (I would kind of consider music pack support to be crossing that line, but I use the feature, so I won't complain too much about it =p)

**Comment by fx02** [comment]
I'm with fraggle, I've removed limits in cndoom and I think choco should stick with vanilla defaults. Messing too much will turn choco to prboom++. I think focus should be on fixing bugs, making it more vanilla and adding features like video recording or basic launcher. Yes, they are not vanilla features but they don't mess up with vanilla code, just add easier usage.

**Comment by CapnClever** [comment]
I tested Colin Phipps's longtics hack by patching it onto a vanilla executable and running it with a PWAD (d2twid.wad). Recording using `-longtics` was no problem; playback (without typing `-longtics`) had no problems either. Attempting to then playback the demo using an unhacked vanilla executable brought up the message "Demo is from a different game version!". This all sounds perfectly reasonable to me. Perhaps Chocolate Doom would do better not to assume 1.91 on playback, then, and require an explicit gameversion? (Or maybe I'm missing the point, simply looking at it from a demo compatibility standpoint.)

Not that I'm saying the above is definitive proof it should stay unchanged. Likewise, the vvHeretic additions are "bleeding-edge technology" that greatly expand upon what Heretic and Hexen can do with their demos, and while I'm grateful they were decided to be included I was also fully prepared for them to be rejected.

---

I don't understand the want for a dogmatic position regarding vanilla hacks. I believe a better way to approach it is to judge the hack on its merits as a feature, using the evidence that "people have managed to implement this fluidly into vanilla" as a point in its favor and no more. There are some other vanilla hacks floating around as well, let's consider them now:

* What about [this heapsize-raising hack](https://www.doomworld.com/vb/post/1049562), to set the allocation up to 32MB? I think everyone can agree that specifically making a toggle for this hack would be redundant, because Chocolate Doom already allows a user to customize the heapsize far lower or higher than vanilla allowed.
* What about [this bug-fixing hack](https://www.doomworld.com/vb/post/674362) that fixes the ouch face, and the "really need" message, and doubled SFX on fast closing doors? These are, by the hack's own admission, bugs-to-be-fixed, and Chocolate Doom ranks the preservation of gameplay behavior as its highest priority (quoting the philosophy document, such behavior "includes the look, feel and sound"). So I'd say they don't really fit, just as much as a WAD only designed to fix minor details in the original graphics isn't going to be accepted as an IWAD: such changes aren't extending gameplay but instead remaking it into what the author believes is "correct". Chocolate Doom's answer to correctness is "what vanilla does", with exceptions only as noted.

Doom-plus is a complicated case because it's not self-contained. The goal is "limit-raising", and it indeed raises many of the limits in a simple manner. But let's suppose the following:
* Chocolate Doom supports a toggle to "raise limits to Doom-plus"
* Player finds a new way to hit a limit that Doom-plus doesn't reach (let's say solidsegs, since that's known), reaches out to Choco devs
* Choco devs state that Chocolate Doom only accepts raises limits via Doom-plus
* Player complains to Doom-plus devs, Doom-plus is patched to include change
* Chocolate Doom suddenly must update as well or not hold to its claim

Alternatively, Chocolate Doom could implement per-limit raises, though that's a slippery slope towards requests for better raising or raising other factors or outright removals and suddenly the settings and the code get complicated by an unbounded measure.

In conclusion, we should treat all requests equally. I'm okay with extensions like longtics and Heretic/Hexen multilevel demos because that's all they do and that's all that can be expected of them, nor do they get in the way of vanilla behavior otherwise. If Chocolate Doom ever hypothetically looked into raising limits, it should do it on its own terms and with its own expectations. Vanilla hacks are not an excuse, only a suggestion.

**Comment by fragglet** [comment]
> `-longtics` came out at the same time period as Doom+. If someone would release a new hack now, in 2016, then you can have doubts. It's been 10 years since there weren't any new hacks. You could trace the line there.

It's better to think of it this way: if `-longtics` came out now, I'd probably be a lot more hesitant to add it to Chocolate Doom. The philosophy of the project was much less clear in the early days of the project, when that feature was added. Sorry, I don't agree that "any hack that was made for vanilla should be emulated" - that's just not my vision for the project.

That said, the advantage of being able to record demos without unduly affecting gameplay is so great that even if Doom v1.91 didn't already exist, I suspect that I would probably have defined an equivalent gameplay format for recording Chocolate Doom demos.

I should probably add a justification for this to the philosophy doc - probably something like "changes are acceptable that allow the player to be able to record the game", but maybe it's possible to express this as a more general principle. I need to give it some thought.

**Comment by AXDOOMER** [comment]
> As fraggle was mentioning about -longtics, it significantly changes game behavior in a way that would be beneficial to players. Speedrunners appreciate the game controlling exactly the same whether recording a demo or not.

Doom+ isn't beneficial to players? I appreciated the fact that I was able to play huge levels using the Doom+ patched vanilla engine. 

> I am not aware of any mappers that actually target Doom-Plus's expanded limits, instead it's either vanilla or limit-removing. 

You're right not one uses these limits as a reference for mapping, but Doom+ rarely crashes on huge levels and I'm glad that this hack exists. At the time I was using Doom+, I generated a huge (extreme size) completely outdoor map using Oblige (huge enough that it was the kind of level that takes two hours to complete). Doom+ didn't crash until I was at the highest point in the level in a corner of the map and looking at the rest of it. This happened on only one map out of the couple of hundreds of Oblige maps that I have generated in my lifetime. So chances that a limit-removing map will crash Doom+ are very low. Solidsegs may cause a crash before those limits are reached. 

> What if your favorite map uses 1060 visplanes on your favorite level? 

According to the point that I've made, it's rare. From my experience, you have do design a map on purpose to reach these limits. Nuts.wad is a great example (more than 1024 monster sprites on screen at once). The Doom+ hack would make a huge number of levels available for Choco-Doomers. 

>  Why not use Crispy Doom for these? It's quite a good fork.

It's not vanilla, not suitable for purists. 

> Where would we draw the line? :)

Like I said, there haven't been any new hacks for a decade, so let's draw the line there. 

> longtics makes the line blurry, 

Longtics seem to be more like a compatibility thing with demos that an emulation of this hack. This hack looks for demos with "lt" appended to their name inside the IWAD or PWAD, something that Chocolate-Doom doesn't do. Fraggle seems to want Chocolate-Doom to be fully compatible with vanilla rather than an emulation. He supports the autodetection of longtics demos for convenience, vanilla didn't do this. It's Chocolate-Doom's goal that has become blurry to me. 

>  Messing too much will turn choco to prboom++.

Chocolate-Doom is already being messed up with a lot lately. V1.2 support will definitively make it look like PrBoom+. Chocolate-Doom will be "PrBoom+ without high resolution support and no options to tweak it". 

> Yes, they are not vanilla features but they don't mess up with vanilla code, just add easier usage.

The Doom+ hack would make it easier to use too. Right now, I have to recompile my Chocolate-Doom-Plus port that incorporates the hack each time there is a change to Chocolate-Doom to keep it up-to-date. Without Doom+ support in Choco, I have to go back to Doom+ in DosBox to play limit removing levels. 

> What about this bug-fixing hack that fixes the ouch face, and the "really need" message, and doubled SFX on fast closing doors?

Do you have a working download link? I've been searching for this hack for a long time.

Personnaly, I would add that hack to Choco, for the same reasons that I support the addition of the Doom+ hack. Since there were not any new hacks for a decade now (9 years exactly), then Doom+ and this hack would be the last ones added to Choco. 

> Player complains to Doom-plus devs, Doom-plus is patched to include change

This hack is out since 10 years. Who's going to update it now? It's a decade, a good place that trace a line and say "we will not accept any new hacks". 

>  That said, the advantage of being able to record demos without unduly affecting gameplay is so great that even if Doom v1.91 didn't already exist, I suspect that I would probably have defined an equivalent gameplay format for recording Chocolate Doom demos.

Doom+ also brings a huge advantage. Limits like Visplanes, Vissprites and Drawsegs are 8 times higher than the unpatched executable, which means that a lot more levels are playable. Lineanims is also an important one and it's 256 times bigger.

**Comment by fragglet** [comment]
@AXDOOMER - if I can summarize your last comment as I read it: you want features Chocolate Doom doesn't have, you don't want features Chocolate Doom does or will soon have; you want to draw a line but it's a different line to the one I want to draw. All of this to me is evidence that you have a different vision or philosophy of what you want in a source port.

And that's great! I'd suggest you make that vision a reality - you already have your chocolate-doom-plus fork, which you can easily build upon and expand. The difference between the two philosophies isn't so great that it will be hard to keep the two in sync. Hacking on the Doom source is a lot of fun and I've always encouraged people to make new and interesting forks. Having a clear vision of what you want is (in my opinion) the most important thing, since many Doom source ports don't even have that.

**Comment by AXDOOMER** [comment]
It's not exactly like you say. 

> you want features Chocolate Doom doesn't have

Sure. I want to implement every hack that was created for vanilla, because that's missing. 

> you don't want features Chocolate Doom does or will soon have

I'd have to think about v1.2 support, I'm not sure about this one, because it's going to look to much like PrBoom+, but I suppose I would add support for this. It's wouldn't just be v1.2 demo support, it would be a full Doom v1.2 emulation. 

Chocolate-Doom is meant to emulate vanilla Doom as closely as possible. In the Readme, vanilla is defined as "Vanilla” is defined as: DOS Doom 1.9 (although there are actually multiple “1.9”s)." You may change this definition soon if you're going to implement v1.2 support. 

Things like the auto-detection of demo's version and stuff like this that was added for convenience, I would completely get rid of this. I would keep the longtics hack of course, but I would completely emulate it. Since there won't be any auto-detection, a longtics demo would crash if `-longtics` is not specified. 

> you want to draw a line but it's a different line to the one I want to draw. 

My line is not "maybe if the feature is useful". That's too abstract. 

My line would be to stop adding new hacks that came after the last hack came out, so my line is set at around 2007 (the "bugsfree" hack). It's soon going to be a decade from now that there were not any new hack released. 

> All of this to me is evidence that you have a different vision or philosophy of what you want in a source port.

Yes and I thought Chocolate-Doom shared that philosophy with me, that's why I didn't understand that the Doom+ hack wasn't added. I thought it was just because it was difficult to implement. 

For the time being, Chocolate-Doom Plus will stay like it is right now. I may call my future port that's going to be even closer to vanilla *Vanilla Doom*. Chocolate-Setup will also have to respect vanilla's standards (BTW the F1 help key, I hate it. I was used to press it to go to the warp menu).

**Comment by linguica** [comment]
Not to delve too deep into pedantry, but there is [at least one WAD](https://www.doomworld.com/idgames/levels/doom/Ports/a-c/blorealm) that explicitly lists Doom+ as a "port" it's been tested with.

**Comment by chungy** [comment]
> Chocolate-Doom is meant to emulate vanilla Doom as closely as possible. In the Readme, vanilla is defined as "Vanilla” is defined as: DOS Doom 1.9 (although there are actually multiple “1.9”s)." You may change this definition soon if you're going to implement v1.2 support.

2.2 probably should have had this definition altered, with 1.666, 1.7, 1.8 support added. :-)

**Comment by AXDOOMER** [comment]
Yes, this is what I mean. Make the line clear so people don't get confused. If there is no intention to add support for Doom v1.2 or older versions of Heretic, Hexen and Strife, then make it clear.

**Comment by fabiangreffrath** [comment]
> not suitable for purists.

@AXDOOMER Could you please file an issue in the Crispy Doom tracker and explain which of the non-optional features leaves you with this impression?

**Comment by fragglet** [comment]
> I may call my future port that's going to be even closer to vanilla **Vanilla Doom**

You can name your port whatever you want, of course, but can I humbly request that you find a more original name? It's going to be incredibly confusing if people start referring to "Vanilla Doom" and it's not clear whether they're referring to your port or the original DOS version.

**Comment by AXDOOMER** [comment]
> but can I humbly request that you find a more original name? 

Don't worry, I'm probably never going to do it. 

> explain which of the non-optional features leaves you with this impression?

I'm a purist, so the fact that you have certain features that are non-vanilla in your port, even if they can be disabled, makes it not suitable for me.

**Comment by chungy** [comment]
Doom-Plus extended limits are certainly not vanilla. :)

**Comment by AXDOOMER** [comment]
Although it doesn't fit the exact definition of a "vanilla software", I still consider it vanilla because it doesn't denaturate the original engine. No, these limits are not vanilla, but it's like vanilla ice cream with a few sprinkles on it. It still tastes vanilla. I use Doom+ only when needed. I'm not purist at the point where I wouldn't play new maps from PWADs because they are not vanilla maps or wouldn't play FreeDM because it uses a DEH patch to rename the maps on the automap.

**Comment by haleyjd** [comment]
I still feel an approach taken here where the limits could be manipulated at compile time would alleviate the need to maintain a zillion downstream "pet" forks that do nothing other than allow more visplanes, w/o imposing any new functionality here. It's a compromise that I don't feel has been given full consideration.

**Comment by sneakernets** [comment]
I agree with @haleyjd , if such things are going to be "in every fork" anyway, it might be of the hacker/tinkerer's best interest to make those features easily tweakable. Not only would that be easier for ports, but it would make it easier for those who want to make a "[doom plus](http://prboom-plus.sourceforge.net/doom-plus.features.html)" of their own with the common features they want (512K alloca stack, 32MB Zone heap, increased visplanes, etc).

Nothing's stopping someone from making doom-plus demos, either. :suspect:

**Comment by linguica** [comment]
Even if they're not exposed to someone using a compiled build, I would be strongly in favor of collecting the various [static limits](https://doomwiki.org/wiki/Static_limits) into their own header file, d_limits.h or what have you. This would at the very least make it much, much more convenient for any enterprising hackers to change them.

**Comment by jmickle66666666** [comment]
I agree with the calls to expose the limits in a much more convenient way for people compiling their own builds, especially considering the common practice of using chocolate doom as a basis for new custom ports; trivialising removing the limits will save time and hassle for every developer looking at making their own port.

**Comment by sneakernets** [comment]
The header file approach sounds great, especially if different Doom versions had different rendering limits, which I have been told *is* the case.
